### PR TITLE
#24 find realm name in Host header (if param missing)

### LINKED ISSUE #24: Support reading the realm name from the "Host" header
We should support using the "Host" header to find the "realm" to use, e.g. a URL like "https://token.services.example.org" would automatically use the realm "/services" for token creation.

### PR DESCRIPTION
**Pull request by hjacobs**
Fixes #24

### CODE REVIEW

**Comment by hjacobs** [comment]
@harti2006 @mrandi please review

File: src/test/java/org/zalando/planb/provider/OIDCCreateTokenIT.java [hunk 1]
```diff
@@ -77,6 +77,47 @@ public void createServiceUserToken() {
...
+
+        try {
+            rest.exchange(request, OIDCCreateTokenResponse.class);
+            fail("Request with invalid Host header should have failed.");
+        } catch (HttpClientErrorException ex) {
+            assertThat(ex.getStatusCode()).isEqualTo(HttpStatus.BAD_REQUEST);
+            assertThat(ex.getResponseBodyAsString()).contains("No matching realm found for token.servicesX.example.org");
+        }
+
+        // Host header contains a valid realm name
```
Comment (mrandi) [inline src/test/java/org/zalando/planb/provider/OIDCCreateTokenIT.java]: can we split this in another test method? would be more clear what the test is intended to, if it fail.

File: src/main/java/org/zalando/planb/provider/OIDCController.java [hunk 2]
```diff
@@ -64,16 +64,21 @@
...
     @ResponseBody
-    OIDCCreateTokenResponse createToken(@RequestParam(value = "realm", required = true) String realmName,
+    OIDCCreateTokenResponse createToken(@RequestParam(value = "realm") Optional<String> realmNameParam,
                                         @RequestParam(value = "grant_type", required = true) String grantType,
                                         @RequestParam(value = "username", required = true) String username,
                                         @RequestParam(value = "password", required = true) String password,
                                         @RequestParam(value = "scope") Optional<String> scope,
-                                        @RequestHeader(name = "Authorization") Optional<String> authorization)
+                                        @RequestHeader(name = "Authorization") Optional<String> authorization,
+                                        @RequestHeader(name = "Host") Optional<String> hostHeader)
```
Comment (mrandi) [inline src/main/java/org/zalando/planb/provider/OIDCController.java]: can we set required for the header?

File: src/main/java/org/zalando/planb/provider/OIDCController.java [hunk 3]
```diff
@@ -64,16 +64,21 @@
...
             throws RealmAuthenticationException, RealmAuthorizationException, JOSEException {
         final Metric metric = new Metric(metricRegistry);
         metric.start();
 
+        final String realmName = realmNameParam.orElseGet(() -> realms.findRealmNameInHost(hostHeader
```
Comment (mrandi) [inline src/main/java/org/zalando/planb/provider/OIDCController.java]: I would check here if the hostHeader is null, in order to avoid null pointer in the RealmConfig line 55 (the pattern split line)

File: src/test/java/org/zalando/planb/provider/RealmConfigTest.java [hunk 4]
```diff
@@ -16,4 +20,28 @@ public void testEnsureLeadingSlashWhenMissing() throws Exception {
...
+    @Test
+    public void testFindRealmInHost() {
+        Set<String> realmNames = ImmutableSet.of("/foo", "/bar");
+        assertThat(RealmConfig.findRealmNameInHost(realmNames, "somethingelse")).isEmpty();
+        assertThat(RealmConfig.findRealmNameInHost(realmNames, "foobar")).isEmpty();
+        assertThat(RealmConfig.findRealmNameInHost(realmNames, "xfoox")).isEmpty();
+        assertThat(RealmConfig.findRealmNameInHost(realmNames, "provider.foo.example.org")).contains("/foo");
+        assertThat(RealmConfig.findRealmNameInHost(realmNames, "provider-bar.example.org")).contains("/bar");
+        // we sort the realm names, so it will return the first one if both match
+        assertThat(RealmConfig.findRealmNameInHost(realmNames, "foo.bar")).contains("/bar");
```
Comment (mrandi) [inline src/test/java/org/zalando/planb/provider/RealmConfigTest.java]: and maybe add a test if we decide to avoid the null check on the controller side:

```
assertThat(RealmConfig.findRealmNameInHost(realmNames, null)) -> throw null pointer
```

File: src/main/java/org/zalando/planb/provider/OIDCController.java (hunk 3 above)
Comment (hjacobs) [inline src/main/java/org/zalando/planb/provider/OIDCController.java]: hostHeader is Optional and cannot be null (using `orElseThrow`)

File: src/main/java/org/zalando/planb/provider/OIDCController.java (hunk 2 above)
Comment (hjacobs) [inline src/main/java/org/zalando/planb/provider/OIDCController.java]: * Host header is not required for HTTP 1.0, see https://www.w3.org/Protocols/HTTP/1.0/spec.html#Request-Header
* the Authorization header I will not touch right now (as it's unrelated to my change)

File: src/test/java/org/zalando/planb/provider/RealmConfigTest.java (hunk 4 above)
Comment (hjacobs) [inline src/test/java/org/zalando/planb/provider/RealmConfigTest.java]: added @NotNull annotation to host param

File: src/test/java/org/zalando/planb/provider/OIDCCreateTokenIT.java (hunk 1 above)
Comment (hjacobs) [inline src/test/java/org/zalando/planb/provider/OIDCCreateTokenIT.java]: split done

File: src/main/java/org/zalando/planb/provider/OIDCController.java (hunk 3 above)
Comment (mrandi) [inline src/main/java/org/zalando/planb/provider/OIDCController.java]: :+1: totally right! didn't think about the sequential flow and only concentrate on hostHeader.get(). In case of hostHeader  = null it will end with the first Exception.
